posix/fdpass.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * File descriptor passing
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Ziemowit Leszczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "fdpass.h"
+
+
+/* FIXME: _Alignof is available since C11 */
+#define CMSG_ALIGN(n)       (((n) + _Alignof(struct cmsghdr) - 1) & ~(_Alignof(struct cmsghdr) - 1))
+#define CMSG_SPACE(n)       (CMSG_ALIGN(sizeof(struct cmsghdr)) + CMSG_ALIGN(n))
+#define CMSG_LEN(n)         (CMSG_ALIGN(sizeof(struct cmsghdr)) + (n))
+#define CMSG_DATA(c)        ((unsigned char *)((struct cmsghdr *)(c) + 1))
+#define CMSG_FIRSTHDR(d, l) ((l) < sizeof(struct cmsghdr) ? (struct cmsghdr *)NULL : (struct cmsghdr *)(d))
+
+#define CMSG_NXTHDR(d, l, c) \
+	({ \
+		struct cmsghdr *_c = (struct cmsghdr *)(c); \
+		char *_n = (char *)_c + CMSG_SPACE(_c->cmsg_len); \
+		char *_e = (char *)(d) + (l); \
+		(_n > _e ? (struct cmsghdr *)NULL : (struct cmsghdr *)_n); \
+	})
+
+
+#define FDPACK_PUSH(p, of, fl) \
+	do { \
+		(p)->fd[(p)->first + (p)->cnt].file = of; \
+		(p)->fd[(p)->first + (p)->cnt].flags = fl; \
+		++(p)->cnt; \
+	} while (0)
+
+#define FDPACK_POP_FILE(p, of) \
+	do { \
+		(of) = (p)->fd[(p)->first].file; \
+		++(p)->first; \
+		--(p)->cnt; \
+	} while (0)
+
+#define FDPACK_POP_FILE_AND_FLAGS(p, of, fl) \
+	do { \
+		(of) = (p)->fd[(p)->first].file; \
+		(fl) = (p)->fd[(p)->first].flags; \
+		++(p)->first; \
+		--(p)->cnt; \
+	} while (0)
```

Comment:
Why macros, not functions?

Reply:
No particular reason, CMSG macros are based on libphoenix macros.

Reply:
Ok, but this is kernel, and this code isn't shared with libphoenix (maybe it should be?). No need for being posixy here and we prefer functions to macros, if possible.